File: src/lib/interface/article.ts
import { ArticleData } from "./article-data";
import { EntityType } from "./entity-type";

export interface ArticleUpdate<E extends ArticleData> {
    id: number;
    entity_type: EntityType;
    title: string | null;
    body: string | null;
    entity: E | null;
}

export interface ArticleInfoResponse {
    id: number;
    entity_type: EntityType;
    title: string;
}

export interface ArticleResponse<E extends ArticleData>
    extends ArticleInfoResponse {
    entity: E;
    body: string;
}

File: src-tauri/src/schema/article.rs
use serde::{Deserialize, Serialize};

use crate::types::EntityType;

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct ArticleInfoSchema {
    pub id: i32,
    pub folder_id: i32,
    pub entity_type: EntityType,
    pub title: String,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct ArticleCreateSchema<E> {
    pub folder_id: i32,
    pub title: String,
    pub data: E,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct ArticleResponseSchema {
    pub id: i32,
    pub folder_id: i32,
    pub entity_type: EntityType,
    pub title: String,
    pub body: String,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct ArticlePropertiesResponseSchema<E> {
    pub id: i32,
    pub folder_id: i32,
    pub entity_type: EntityType,
    pub title: String,
    pub entity: Option<E>,
}
